src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,997 @@
+  // it
+  U new_zeros = bits._zeros | (match_mask & ~bounds._lo);
+  U new_ones = bits._ones | (match_mask & bounds._lo);
+  bool progress = (new_zeros != bits._zeros) || (new_ones != bits._ones);
+  bool present = ((new_zeros & new_ones) == U(0));
+  return {progress, present, {new_zeros, new_ones}};
+}
+
+// Try to tighten both the bounds and the bits at the same time.
+// Iteratively tighten one using the other until no progress is made.
+// This function converges because at each iteration, some bits that are unknown
+// are made known. As there are at most 64 bits, the number of iterations should
+// not be larger than 64.
+// This function is called simple because it deals with a simple intervals (see
+// TypeInt at type.hpp).
+template <class U>
+static SimpleCanonicalResult<U>
+canonicalize_constraints_simple(const RangeInt<U>& bounds, const KnownBits<U>& bits) {
+  AdjustResult<KnownBits<U>> canonicalized_bits = adjust_bits_from_bounds(bits, bounds);
+  if (canonicalized_bits.empty()) {
+    return SimpleCanonicalResult<U>::make_empty();
+  }
+  AdjustResult<RangeInt<U>> canonicalized_bounds{true, true, bounds};
+  // Since bits are derived from bounds in the previous iteration and vice
+  // versa, if one does not show progress, the other will also not show
+  // progress, so we terminate early
+  while (true) {
+    canonicalized_bounds = adjust_bounds_from_bits(canonicalized_bounds._result, canonicalized_bits._result);
+    if (!canonicalized_bounds._progress || canonicalized_bounds.empty()) {
+      return SimpleCanonicalResult<U>(canonicalized_bounds._present, canonicalized_bounds._result, canonicalized_bits._result);
+    }
+    canonicalized_bits = adjust_bits_from_bounds(canonicalized_bits._result, canonicalized_bounds._result);
+    if (!canonicalized_bits._progress || canonicalized_bits.empty()) {
+      return SimpleCanonicalResult<U>(canonicalized_bits._present, canonicalized_bounds._result, canonicalized_bits._result);
+    }
+  }
+}
+
+// Tighten all constraints of a TypeIntPrototype to its canonical form.
+// i.e the result represents the same set as the input, each bound belongs to
+// the set and for each bit position that is not constrained, there exists 2
+// values with the bit value at that position being set and unset, respectively,
+// such that both belong to the set represented by the constraints.
+template <class S, class U>
+typename TypeIntPrototype<S, U>::CanonicalizedTypeIntPrototype
+TypeIntPrototype<S, U>::canonicalize_constraints() const {
+  RangeInt<S> srange = _srange;
+  RangeInt<U> urange = _urange;
+  // Trivial contradictions
+  if (srange._lo > srange._hi ||
+      urange._lo > urange._hi ||
+      (_bits._zeros & _bits._ones) != U(0)) {
+    return CanonicalizedTypeIntPrototype::make_empty();
+  }
+
+  // Trivially canonicalize the bounds so that srange._lo and urange._hi are
+  // both < 0 or >= 0. The same for srange._hi and urange._ulo. See TypeInt for
+  // detailed explanation.
```

Comment:
You seem to suggest that `urange._hi` can be `< 0`. But it is unsigned, so that is confusing.
This looks like Lemma 3 from TypeInt is relevant here, right? You might want to state that here,
and still also bring this ASCII art up again here:
```
 *     Signed:
 *     -----lo=========uhi---------0--------ulo==========hi-----
 *     Unsigned:
 *                                 0--------ulo==========hi----------lo=========uhi---------
```
Also, does `S(urange._lo) > S(urange._hi)` imply `U(srange._lo) > U(srange._hi)`?
- If yes, assert it!
- If no: are we missing an optimization?

Reply:
It could also be that I'm missing some things here... So it could be good to refer to the exact things in `TypeInt` and the Lemma numbers you are using here. Otherwise it is a lot of work for the reader to check what you are doing here.

Reply:
Here the bounds may take any value, so we cannot have `S(urange._lo) > S(urange._hi)` imply `U(srange._lo) > U(srange._hi)` yet. I hope the description I have done is now clearer.

Reply:
Here the bounds may take any value, so we cannot have S(urange._lo) > S(urange._hi) imply U(srange._lo) > U(srange._hi) yet. I hope the description I have done is now clearer.